src/internal/animate.ts

```diff
@@ -0,0 +1,110 @@
+  return rects;
+}
+
+/*
+  Animate DOM elements based on the FLIP technique
+  - https://aerotwist.com/blog/flip-your-animations/
+  - https://css-tricks.com/animating-layouts-with-the-flip-technique/
+
+  This can be useful when the initial dimensions or position of the element is not known,
+  so the initial offset or scaling needs to be retrieved via JS.
+
+  Caveat: this currently does not support elements having CSS transforms in the end state.
+  These would be overridden in the animation instead of combined.
+ */
+export function animate({
+  oldState,
+  elements,
+  onTransitionsEnd,
+  newElementInitialState,
+}: {
+  elements: Record<string | number, HTMLElement | null>;
+  oldState: Record<string | number, DOMRect>;
+  onTransitionsEnd?: () => void;
+  newElementInitialState?: (newRect: DOMRect) => { scale?: number; y?: number };
+}) {
+  // First, apply the transform that will make the elements "look like" in the start position
+  for (const id in elements) {
+    const element = elements[id];
+    const oldRect = oldState[id];
+    if (element) {
+      const newRect = element.getBoundingClientRect();
+      const noOpTransform = { scale: 1, x: 0, y: 0 };
+      // Calculate initial position.
+      // If the element didn't exist previously, use the newElementInitialState function if provided.
+      // If not, default to no transitions (scale: 1, y: 0)
+      const calculatedInverseTransform = oldRect
+        ? {
+            scale: oldRect.width / newRect.width,
+            x: (oldRect.left + oldRect.right) / 2 - (newRect.left + newRect.right) / 2,
+            y: (oldRect.top + oldRect.bottom) / 2 - (newRect.top + newRect.bottom) / 2,
+          }
+        : newElementInitialState
+        ? newElementInitialState(newRect)
+        : {};
+      const inverseTransform = { ...noOpTransform, ...calculatedInverseTransform };
+      // Apply this initial change, without animating
+      element.style.transitionProperty = 'none';
+      element.style.transform = `scale(${inverseTransform.scale}) translate(${inverseTransform.x}px, ${inverseTransform.y}px)`;
+      if (!oldRect) {
+        // If the element didn't exist, then fade it in
+        // (besides any other possibly defined transitions based on `newElementInitialState`)
+        element.style.opacity = '0';
+      }
+    }
+  }
+
+  // Animate from the initial state to the end state
+  requestAnimationFrame(() => {
+    const ongoingAnimations = new Set();
+    for (const id in elements) {
+      const element = elements[id];
+      if (element) {
+        const oldRect = oldState[id];
+        if (oldRect) {
+          // Animate from here on
+          element.style.transitionProperty = `transform`;
+          // Unset inline CSS transforms so that the final state is applied
+          element.style.transform = '';
+        } else {
+          // If the element didn't exist previously, fade in as well
+          element.style.transitionProperty = `transform, opacity`;
```

Comment:
The reason for this helper to exist is for animations in cases where the initial or final position might be in arbitrary places, so you need some JavaScript magic to calculate the value of the `transform` property. In general cases, for animating opacity you don't need all this because you have the static initial and end state so you can do it in a simpler way, e.g, with just CSS. The only reason we animate opacity it here is for the special case when the element didn't exist in the initial state, in which case I considered it makes sense as a general case to fade it in (but the animation properties such as duration and curve are still defined in CSS).

I guess it's unlikely that opacity conflicts with something else (CSS), but I prefer to keep these JS DOM manipulations minimal just in case.

Reply:
Got it, thank you for the explanation!